### PR TITLE
Hash cache with 0 salt

### PR DESCRIPTION
**Pull request by shashi**
caches in more cases than #171

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -296,10 +296,11 @@ arguments(x::Term) = getfield(x, :arguments)
 hashvec(xs, z) = foldr(hash, xs, init=z)
 
 function Base.hash(t::Term{T}, salt::UInt) where {T}
+    !iszero(salt) && hash(hash(t, zero(UInt)), salt)
```
Comment (YingboMa) [inline src/types.jl:299]: Shouldn't there be a return?

Comment (YingboMa) [reply src/types.jl:299]:
```suggestion
    !iszero(salt) && return hash(hash(t, zero(UInt)), salt)
```

Comment (shashi) [reply src/types.jl:299]: yup

File: src/types.jl [hunk 2]
```diff
@@ -296,10 +296,11 @@ arguments(x::Term) = getfield(x, :arguments)
...
     h = t.hash[]
-    (iszero(salt) && !iszero(h)) && return h
+    !iszero(h) && return h
     h′ = hashvec(arguments(t), hash(operation(t), hash(T, salt)))
-    iszero(salt) && (t.hash[] = h′)
+    (t.hash[] = h′)
```
Comment (YingboMa) [inline src/types.jl:303]:
```suggestion
    t.hash[] = h′
```

**Comment by shashi** [comment]
MTK fails looking weird.

**Comment by YingboMa** [comment]
No... Tests fails with this PR should probably don't merge it.

**Comment by YingboMa** [comment]
It was passing before

**Pull request merged**